### ISSUE TITLE
Linting fails during compilation of ink! 3.4.0 contracts

### ISSUE THREAD

**Issue by Raid5594** [paritytech/cargo-contract#1381]
**Describe the bug**
During linting step of compilation, the build fails. Please see attached the screenshot with the error.

**To Reproduce**
1. cargo install cargo-dylint dylint-link
2. cargo install cargo-contract --version 1.5.0 --force --locked
3. cargo contract build

**Expected behavior**
Build not to fail :)

**Screenshots**
<img width="1457" alt="Screenshot 2023-10-25 at 18 43 35" src="https://github.com/paritytech/cargo-contract/assets/52794079/6e74262c-26f7-4a82-9e62-0f1096b40ce3">

**Environment (please complete the following information):**
 - OS: MacOS M2
 - `cargo-contract` version 1.5.0
 - ink! version: 3.4.0

**Additional context**
What is peculiar, is that we tested the process with different dylint versions, and on some local machines cargo-dylint@2.1.10 works with flow, how neither on my machine, no within Github Actions we were not able to reproduce the behaviour.

**Comment by ascjones** [comment]
The error output says you need a newer version of Rust, `1.65` or newer.

**Comment by jubnzv** [comment]
That's right, the `regex-syntax` which is an indirect dependency of [`assert_cmd`](https://github.com/paritytech/cargo-contract/blob/6331b9cb8f14b2771b49df910db4eb134d08dc03/crates/cargo-contract/Cargo.toml#L56) requires Rust >=1.65.

@Raid5594 Is it possible to update the Rust toolchain on your machines?

**Comment by Raid5594** [comment]
@ascjones but I'm using rust 1.70.0 for this build, 
I can rerun with explicit toolchain selection to show that it doesn't resolve the error

**Comment by smiasojed** [comment]
The toolchain for linting is hardcoded in `cargo-contract/ink_linting/rust-toolchain.toml`:
```
channel = "nightly-2022-06-30"
```
You can update it, but you will likely encounter compatibility issues with crates.

**Comment by Raid5594** [comment]
@smiasojed thanks for spotting this. But why is it not compatible with the corresponding cargo-contract version in the first place?

**Comment by smiasojed** [comment]
@Raid5594 In this ticket, you pointed out that this setup is working on one of the machines. Could you take a look at the Cargo.lock for the contract on the machine where it is working and compare it to the machines where it is not working?
I suspect that in cases where it is not working, some crates have been updated.

**Comment by jubnzv** [comment]
@ Raid5594 Could you please provide a link to the Github Actions workflow failing with this error?
I'm still not able to reproduce this locally, so I need the exact versions of toolchain and libraries.

**Comment by smiasojed** [comment]
@Raid5594 I reproduced it locally. It is an issue with dylint compilation using this old toolchain.
Please use: cargo install cargo-dylint dylint-link --locked --version  2.0.10

**Comment by smiasojed** [comment]
Feel free to reopen it if the issue still exists on your end.
